File: cpp/src/order_book.cpp
#include "order_book.h"
#include <chrono>
#include <algorithm>

OrderBook::OrderBook() {}

void OrderBook::add_order(Order& order) {
    order.timestamp = std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();

    if (order.side == Side::Buy) {
        auto& price_level = bids[order.price];
        price_level.total_quantity += order.quantity;
        price_level.orders.push_back(order);
    } else {
        auto& price_level = asks[order.price];
        price_level.total_quantity += order.quantity;
        price_level.orders.push_back(order);
    }
}

bool OrderBook::cancel_order(uint64_t order_id, Side side, double price) {
    auto* book_side = (side == Side::Buy) ? &bids : (std::map<double, PriceLevel, std::greater<double>>*) &asks;

    if (side == Side::Buy) {
        auto it = bids.find(price);
        if (it == bids.end()) {
            return false;
        }
        auto& price_level = it->second;
        auto& orders = price_level.orders;
        auto order_it = std::find_if(orders.begin(), orders.end(), 
                                     [order_id](const Order& o) { return o.order_id == order_id; });

        if (order_it != orders.end()) {
            price_level.total_quantity -= order_it->quantity;
            orders.erase(order_it);
            if (orders.empty()) {
                bids.erase(it);
            }
            return true;
        }
    } else {
        auto it = asks.find(price);
        if (it == asks.end()) {
            return false;
        }
        auto& price_level = it->second;
        auto& orders = price_level.orders;
        auto order_it = std::find_if(orders.begin(), orders.end(), 
                                     [order_id](const Order& o) { return o.order_id == order_id; });

        if (order_it != orders.end()) {
            price_level.total_quantity -= order_it->quantity;
            orders.erase(order_it);
            if (orders.empty()) {
                asks.erase(it);
            }
            return true;
        }
    }

    return false;
}

bool OrderBook::modify_order(uint64_t order_id, Side side, double price, uint64_t new_quantity) {
    auto* book_side = (side == Side::Buy) ? &bids : (std::map<double, PriceLevel, std::greater<double>>*) &asks;

    if (side == Side::Buy) {
        auto it = bids.find(price);
        if (it != bids.end()) {
            auto& price_level = it->second;
            for (auto& order : price_level.orders) {
                if (order.order_id == order_id) {
                    price_level.total_quantity -= order.quantity;
                    order.quantity = new_quantity;
                    price_level.total_quantity += new_quantity;
                    return true;
                }
            }
        }
    } else {
        auto it = asks.find(price);
        if (it != asks.end()) {
            auto& price_level = it->second;
            for (auto& order : price_level.orders) {
                if (order.order_id == order_id) {
                    price_level.total_quantity -= order.quantity;
                    order.quantity = new_quantity;
                    price_level.total_quantity += new_quantity;
                    return true;
                }
            }
        }
    }
    return false;
}

std::optional<double> OrderBook::get_best_bid() const {
    if (bids.empty()) {
        return std::nullopt;
    }
    return bids.begin()->first;
}

std::optional<double> OrderBook::get_best_ask() const {
    if (asks.empty()) {
        return std::nullopt;
    }
    return asks.begin()->first;
}

File: rust/src/order_book.rs
//! Represents the order book.
use std::collections::{BTreeMap, VecDeque};
use crate::order::{Order, Side};

pub struct PriceLevel {
    pub total_quantity: u64,
    pub orders: VecDeque<Order>,
}

impl PriceLevel {
    pub fn new() -> Self {
        PriceLevel {
            total_quantity: 0,
            orders: VecDeque::new(),
        }
    }
}

pub struct OrderBook {
    bids: BTreeMap<u64, PriceLevel>,
    asks: BTreeMap<u64, PriceLevel>,
}

impl OrderBook {
    pub fn new() -> Self {
        OrderBook {
            bids: BTreeMap::new(),
            asks: BTreeMap::new(),
        }
    }

    pub fn add_order(&mut self, mut order: Order) {
        order.timestamp = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos() as u64;

        let (book_side, price) = match order.side {
            Side::Buy => (&mut self.bids, order.price),
            Side::Sell => (&mut self.asks, order.price),
        };

        // ...existing code...
        let price_level = book_side.entry(price).or_insert_with(PriceLevel::new);
        price_level.total_quantity += order.quantity;
        price_level.orders.push_back(order);
    }

    pub fn cancel_order(&mut self, order_id: u64, side: Side, price: u64) -> bool {
        let book_side = match side {
            Side::Buy => &mut self.bids,
            Side::Sell => &mut self.asks,
        };

        if let Some(price_level) = book_side.get_mut(&price) {
            if let Some(index) = price_level.orders.iter().position(|o| o.order_id == order_id) {
                let order = price_level.orders.remove(index).unwrap();
                price_level.total_quantity -= order.quantity;
                if price_level.orders.is_empty() {
                    book_side.remove(&price);
                }
                return true;
            }
        }
        false
    }

    pub fn modify_order(&mut self, order_id: u64, side: Side, price: u64, new_quantity: u64) -> bool {
        let book_side = match side {
            Side::Buy => &mut self.bids,
            Side::Sell => &mut self.asks,
        };

        if let Some(price_level) = book_side.get_mut(&price) {
            for order in &mut price_level.orders {
                if order.order_id == order_id {
                    price_level.total_quantity -= order.quantity;
                    order.quantity = new_quantity;
                    price_level.total_quantity += new_quantity;
                    return true;
                }
            }
        }
        false
    }
}